src/pipeline/src/etl.rs

```diff
@@ -97,23 +223,178 @@ impl<T> Pipeline<T>
+        match val {
+            serde_json::Value::Object(map) => {
+                let mut index = 0;
+
+                // because of the key in the required_keys is ordered
+                for (payload_key, payload_value) in map.into_iter() {
+                    // find the key in the required_keys
+                    let mut current_index = index;
+                    while current_index < self.required_keys.len() {
+                        if self.required_keys[current_index] == payload_key {
+                            result[current_index] = payload_value.try_into()?;
+                            index += 1;
+                            break;
+                            // can not find the key in the required_keys
+                        } else {
+                            // find the key in the required_keys
+                            current_index += 1;
+                        }
+                    }
+                }
+            }
+            serde_json::Value::String(_) => {
+                result[0] = val.try_into()?;
+            }
+            _ => {
+                return Err("expect object".to_string());
+            }
+        }
+        Ok(())
+    }
+
+    pub fn init_intermediate_state(&self) -> Vec<Value> {
+        vec![Value::Null; self.intermediate_keys.len()]
+    }
+
+    pub fn reset_intermediate_state(&self, result: &mut [Value]) {
+        for i in result {
+            *i = Value::Null;
+        }
+    }
+
+    pub fn processors(&self) -> &processor::Processors {
+        &self.processors
+    }
+
+    pub fn transformer(&self) -> &T {
+        &self.transformer
+    }
 
-        self.transformer.transform(val)
+    /// Required fields in user-supplied data
+    pub fn required_keys(&self) -> &Vec<String> {
+        &self.required_keys
+    }
+
+    /// All output keys from the pipeline
+    pub fn output_keys(&self) -> &Vec<String> {
+        &self.output_keys
+    }
+
+    /// intermediate keys from the processors
+    pub fn intermediate_keys(&self) -> &Vec<String> {
+        &self.intermediate_keys
+    }
+
+    pub fn schemas(&self) -> &Vec<greptime_proto::v1::ColumnSchema> {
+        self.transformer.schemas()
     }
 }
 
 #[cfg(test)]
 mod tests {
 
+    use greptime_proto::v1::value::ValueData;
     use greptime_proto::v1::{self, ColumnDataType, SemanticType};
 
     use crate::etl::transform::GreptimeTransformer;
     use crate::etl::{parse, Content, Pipeline};
+    use crate::Value;
+
+    #[test]
+    fn test_pipeline_prepare() {
+        let input_value_str = r#"
+            {
+                "my_field": "1,2",
+                "foo": "bar"
+            }
+        "#;
+        let input_value: serde_json::Value = serde_json::from_str(input_value_str).unwrap();
+
+        let pipeline_yaml = r#"
+---
+description: Pipeline for Apache Tomcat
+
+processors:
+  - csv:
+      field: my_field, my_field,field1, field2
```

Comment:
What a `field`... I feel the same way as you. @paomian

Reply:
In the next step. We can add a new yaml field `target_fields` to specify the target_fields field.